modepy/spaces.py

```diff
@@ -0,0 +1,146 @@
+"""
+.. currentmodule:: modepy
+
+Function Spaces
+---------------
+
+.. autoclass:: FunctionSpace
+.. autoclass:: PN
+.. autoclass:: QN
+.. autoclass:: space_for_shape
+
+Redirections to Canonical Names
+-------------------------------
+
+.. currentmodule:: modepy.spaces
+
+.. class:: FunctionSpace
+
+    See :class:`modepy.FunctionSpace`.
+
+.. class:: PN
+
+    See :class:`modepy.PN`.
+
+.. class:: QN
+
+    See :class:`modepy.QN`.
+"""
+
+__copyright__ = "Copyright (C) 2020 Andreas Kloeckner"
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+
+from functools import singledispatch
+from modepy.shapes import Shape, Simplex, Hypercube
+
+
+# {{{ function spaces
+
+class FunctionSpace:
+    r"""An opaque object representing a finite-dimensional function space
+    of functions :math:`\mathbb R^n \to :math:`\mathbb R`.
+
+    .. attribute:: spatial_dim
+
+        :math:`n` in the above definition, the number of spatial dimensions
+        in which the functions in the space operate.
+
+    .. attribute:: space_dim
```

Comment:
TBH, I hate both of these names. I picked them to be obnoxiously similar because I wanted to sow enough confusion that people would stumble and realize the difference. `dim` IMO is bad because it could easily/justifiably be either.

But I'm open to other suggestions.

Reply:
Yeah, I think I just convinced myself that it makes more sense to call `spatial_dim` just `dim`, since it's used like that everywhere else in the code.

Can't really think of anything nicer either, I'm afraid.